### ISSUE TITLE
Improve documentation for `typescript/no-unused-vars`

### ISSUE THREAD

**Issue by OliverJAsh** [nzakas/eslint-plugin-typescript#33]
There are several use cases `typescript/no-unused-vars` does not yet support, such as `type` declarations:

``` ts
type Foo = string;
// unexpected error
const foo: Foo = 'foo';
```

Would it be possible to track the status of this somewhere in the documentation? The docs for `typescript/no-unused-vars` do not make any mention of this.

I would be happy to raise a PR, but I don't know the full details of what is and isn't supported.

**Comment by weirdpattern** [comment]
@JamesHenry We can close this issue now 😃

**Comment by OliverJAsh** [comment]
Actually I don't think we can? My PR added this rule to the list, but it doesn't add more detail as per this issue.

**Comment by weirdpattern** [comment]
@JamesHenry can we consider removing this rule in favor of TypeScript own flags [--noUnusedParameters and --noUnusedLocals]
(https://github.com/Microsoft/TypeScript/wiki/What%27s-new-in-TypeScript#flag-unused-declarations-with---nounusedparameters-and---nounusedlocals)?.

I think this rule is causing a lot of confusion and giving a lot of false/positives. 
If we already get this out of box, then I don't see the point in keeping it... thoughts?

**Comment by mightyiam** [comment]
I think that TypeScript did not include `noUnusedParameters` and `noUnusedLocals` in `strict` because they see them more as style things. I think they might even regret having these options at all, as they are supposed to be linting issues and not type issues.

**Comment by lostfictions** [comment]
I appreciate having a `no-unused-vars` linter rule because I may not necessarily want `tsc` to return an error (one that might, say, fail a CI build) on something that I might consider more of a warning. (And there's unfortunately no way to configure the severity of the TypeScript flags.)

**Comment by weirdpattern** [comment]
This is a very good point. I will continue working on this rule for as long as it's needed.

**Comment by tenshiemi** [comment]
What is the correct setting for this rule? If I'm understanding correctly, it should prevent ESLint from erroneously throwing these errors, but I can't for the life of me to get it to work and since it's not operating as a rule so much as a workaround it's not clear to me what value I should set it to.

**Comment by weirdpattern** [comment]
Hi @tenshiemi, this rule extends (eslint/no-unused-vars)[https://eslint.org/docs/rules/no-unused-vars] to include TypeScript syntax. 

You will need both defined in your eslint config to make it work. What I usually do is extend the eslint:recommended settings, then manually enable typescript/no-unused-vars

Hope this helps.

**Comment by tenshiemi** [comment]
This is what I have both it throws a bunch of no-unused-vars and no-undef warnings:
```
{
    "parser": "typescript-eslint-parser",
    "plugins": [
        "react",
        "standard",
        "typescript",
        "prettier"
    ],
    "extends": [
        "eslint:recommended",
        "plugin:react/recommended",
        "prettier",
        "prettier/react",
        "prettier/standard"
    ],
    "parserOptions": {
        "sourceType": "module",
        "ecmaFeatures": {
            "jsx": true
        }
    },
    "rules": {
        "no-undef": "warn",
        "no-unused-vars": "warn",
        "typescript/no-use-before-define": "warn",
        "typescript/no-unused-vars": "warn",
        "prettier/prettier": ["error", {
            "parser": "typescript",
            "semi": true,
            "singleQuote": true,
            "tabWidth": 4,
            "trailingComma": "all"
        }]
    }
```

**Comment by weirdpattern** [comment]
@tenshiemi, we are still working on this rule, we keep finding scenarios that we missed and have to be included.
Could you please provide a snippet of the code that is failing?

**Comment by tenshiemi** [comment]
@weirdpattern I just realized it works when I run `arc lint`, just not when I run eslint directly on a file, so that's a relief!

an example is 
` 10:18  warning  'SearchTableState' is not defined                                                                                                  no-undef`
when there is `export interface SearchTableState {
    data: object[];
    searchText: string;
}` in the code

**Comment by aaronjensen** [comment]
Here's a scenario where it fails:

```ts
import { FieldRenderProps } from 'react-final-form'

const Error = ({ meta }: { meta: FieldRenderProps['meta'] }) => null
```

**Comment by macklinu** [comment]
As of the latest commit (5e410a36c43d82ab679bed8a8ef68a655008fde9), I'm seeing this snippet pass as a local `no-unused-vars` test case. Can add it to the codebase in a PR if that would be helpful.

```ts
import { FieldRenderProps } from 'react-final-form'

const Error = ({ meta }: { meta: FieldRenderProps['meta'] }) => null
```

I've confirmed the following as a failing test and will open a PR to address what I believe is the issue:

```ts
import { Foo } from './types';

class Bar<T extends Foo> {}
```

I'm not able to reproduce "'ExecaOptions' is defined by never used" in the following snippet. I see the following errors locally. @milesj do you have a reproducible test case of this issue, perhaps by adding a failing test in a PR? That would be awesome 😄 

```ts
import execa, { Options as ExecaOptions } from 'execa';

function foo(options: ExecaOptions) {}
```

```
AssertionError [ERR_ASSERTION]: Should have no errors but had 3: 
[ { ruleId: 'no-unused-vars',
    severity: 1,
    message: '\'execa\' is defined but never used.',
    line: 2,
    column: 8,
    nodeType: 'Identifier',
    source: 'import execa, { Options as ExecaOptions } from \'execa\';',
    endLine: 2,
    endColumn: 13 },
  { ruleId: 'no-unused-vars',
    severity: 1,
    message: '\'foo\' is defined but never used.',
    line: 4,
    column: 10,
    nodeType: 'Identifier',
    source: 'function foo(options: ExecaOptions) {}',
    endLine: 4,
    endColumn: 13 },
  { ruleId: 'no-unused-vars',
    severity: 1,
    message: '\'options\' is defined but never used.',
    line: 4,
    column: 14,
    nodeType: 'Identifier',
    source: 'function foo(options: ExecaOptions) {}',
    endLine: 4,
    endColumn: 35 } ]
```

**Comment by milesj** [comment]
@macklinu Here's the execa full code: https://github.com/milesj/boost/blob/master/src/Routine.ts#L111 It's probably the wrapping parens + union causing it.

**Comment by milesj** [comment]
Think I found a few more.

```js
import { Foo, Bar } from './types';

// Doesnt catch either mapped types 
const NAMES: { [name in Foo]: Bar } = {};
```

**Comment by macklinu** [comment]
@milesj pushed up 9f0caf3 in #135 to address https://github.com/nzakas/eslint-plugin-typescript/issues/33#issuecomment-400063994.

**Comment by milesj** [comment]
Any update on this? If you're open to assigning collaborators, I could help go through some of these issues/PRs.

**Comment by Radivarig** [comment]
```js
// warning 'B' is defined but never used
const a: Array<{b: B}> = [] 
```
